### ISSUE TITLE
Manually install the compat version of the CUDA toolkit in the container

### ISSUE THREAD

**Issue by remram44** [selkies-project/docker-nvidia-glx-desktop#44]
Currently there are only 4 tags `18.04`, `20.04`, `22.04`, `latest`. Whenever you release a new version, you update every tag and the previous images are gone.

It would be nice to be able to keep referencing a specific version or to grab an old release. For example you could keep `22.04-20230801` (never updated) beside `22.04` (updated).

Case in point: it is impossible to get any Xfce images since every tag has been overwritten since [the switch to KDE](https://github.com/selkies-project/docker-nvidia-glx-desktop/commit/0bad542b55afc196009f9fad5b7d15175d7b9d11)

**Comment by ehfd** [comment]
There were issues with the Xfce images affecting usage (a pretty peculiar bug that looked like simple mistakes from the dev that wasn't patched and just blacked out everything at startup) that made me discourage maintaining them. And critical issues in the MATE images beforehand as well (issue with NVIDIA drivers > 490 specific to the DE not patched in Focal), related to the DE too. Therefore, I cannot accept any maintenance requests for Xfce, and this was why it was removed.

KDE isn't perfect, but at least it's maintainable. (And this container is for VR as well so I had to choose between GNOME and KDE which supported XRDesktop  - of which GNOME doesn't work without systemd).

**Comment by ehfd** [comment]
(The issue with NVIDIA's 535 driver right now is with Xorg, which is DE-independent.)

**Comment by remram44** [comment]
Oh I'm not saying you should maintain them, just offer a way to find them, like "normal" release artifacts would be findable.

**Comment by ehfd** [comment]
One more thing; potential (high) security risk in years-old containers. Better to check out a commit you would like to be on git then build the Dockerfile manually for the latest security updates.

**Comment by ehfd** [comment]
May consider if someone joins in as a maintainer. I feel that this is extra work which doesn't align with my everyday container usage and rolling release patterns.

**Comment by remram44** [comment]
I am automatically copying images to [remram44/selkies-docker-nvidia-glx-desktop](https://github.com/remram44/selkies-docker-nvidia-glx-desktop/pkgs/container/selkies-docker-nvidia-glx-desktop/versions) for everyone wanting stable tags.

**Comment by ehfd** [comment]
> Having containers break or suddenly change from Xfce to KDE just because it restarted on a different node and pulled something new is not something I want my users to deal with.

Ah, well yes. I knew it was an issue. Sorry for the inconvenience. But it was our roadmap and the decision to move away from Xfce was made because 1) there was a very uncomfortable bug which caused technical burdens, and 2) there were measurable performance issues despite typical perceptions.

I'll try to leave the built images (that can be accessed from the image hash ID) from now on that you could tag to your fork, but please do understand that 1) GitHub may start nagging our project to remove old images or do it themselves any day (I'm surprised they are just stacking exabytes of container images with no cleanup policy), 2) such breaking changes in latest may happen because I must adapt to new conditions, and 3) old images may stop working suddenly because of new upstream bugs or deprecation.

Users are encouraged to fork when they would like to, as this project is meant to be a template which at the same time tries its best to work out of the box.

**Comment by ehfd** [comment]
tl;dr: If you have a configuration (desktop environment, additional wide range packages) that you use frequently and you want it to be supported, I will do it if you continuously provide bug reports.

I will only implement the "multiple CUDA versions" section in this proposal if there is no such volunteer.
Else, our development direction is probably your best shot, and you should keep your own forked image builds to choose when to accept our breaking changes.

I still do not recommend using our old builds, because who knows what would happen from old ca-bundle and ca-certificate versions, old Firefox versions, and whatever else?...

**Comment by ehfd** [comment]
Key issues I faced with Xfce that you should probably read:

1. When the `~/.cache` or `~/.config/xfce4` is not cleared before container start, for some reason, the desktop environment starts with a black screen. It wasn't fixed in Focal.
2. Xfce actually takes on more disk space and RAM (compared to KDE), primarily because they have programs part of the stack which was borrowed from GNOME, MATE, and other desktop environments. This primarily leads to not being able to share dependencies across programs.
3. Its compositor is intensive and impacts performance. Also, it's hard to disable it in the container, as the home directory must be tinkered. KDE with compositor disabled ([which should be disabled with remote desktops or VR/gaming applications anyway](https://linux-gaming.kwindu.eu/index.php?title=Compositor_(X11))) has better results.
4. Even if it's pretty intensive, it really offers not much. The default theme (which is hard to customize within containers) offers not much capabilities.
5. The number of developers are limited and the stack itself is falling back with time. While Xfce are making heroic efforts to keep maintaining, the dependencies are falling behind compared to GNOME and other GTK+ desktop environments. KDE has multiples of more developers working on the project, and you can go to their IRC channels to ask about capabilities instead of relying on me.

**Comment by ehfd** [comment]
So, the settlement:

1. [Intel's Clear Linux](https://hub.docker.com/_/clearlinux/tags) manages container tags the same way as I do. But I acknowledge that I make breaking changes including going from one desktop environment to another. This breaks some users.

2. I cannot do tagged releases (Debian/Ubuntu model). This substantially slows my development speed that I made, and every tag would is always deprecated by the next one.

3. So, my compromise is the Arch Linux model, where the tags are tagged by the time it was built.

4. In addition, I can also use a more diverse set of CUDA versions.

**Comment by ehfd** [comment]
@remram44 All builds are now also tagged with the time of each commit, with the form `ghcr.io/selkies-project/nvidia-glx-desktop:22.04-20230906134218` and `ghcr.io/selkies-project/nvidia-glx-desktop:20.04-20230906134218`.


Please tell me if this is what you want. Again, thanks for your feedback and helping me resolve repository behaviors into best-practice.

This issue will still be kept open, as I would like to manually install `cuda-compat` in the container. to support all driver versions since 450 while opportunistically upgrading the CUDA version.

**Comment by remram44** [comment]
I'm going to unfollow this issue since it seems to have completely changed topic, from Docker images history to CUDA dylibs. Good luck with CUDA!

**Comment by ehfd** [comment]
@remram44 
One thing. There was a CVE with libwebp and libvpx. If you have images built before Oct 3, you should rebuild it because it was a major security issue.

**Comment by ehfd** [comment]
So far (updating):
Forward compatibility working on: Quadro M6000
Forward compatibility not working on: GTX 1070

**Comment by ehfd** [comment]
My final solution:

```
# Extract NVRTC dependency, https://developer.download.nvidia.com/compute/cuda/redist/cuda_nvrtc/LICENSE.txt
cd /tmp && curl -fsSL -o nvidia_cuda_nvrtc_linux_x86_64.whl "https://developer.download.nvidia.com/compute/redist/nvidia-cuda-nvrtc/nvidia_cuda_nvrtc-11.0.221-cp36-cp36m-linux_x86_64.whl" && unzip -joq -d ./nvrtc nvidia_cuda_nvrtc_linux_x86_64.whl && cd nvrtc && chmod 755 libnvrtc* && find . -maxdepth 1 -type f -name "*libnvrtc.so.*" -exec sh -c 'ln -snf $(basename {}) libnvrtc.so' \; && mv -f libnvrtc* /opt/gstreamer/lib/x86_64-linux-gnu/ && cd /tmp && rm -rf /tmp/*
```

Since Selkies-GStreamer only requires `libnvrtc.so`, I extracted the libraries from a `.whl` file and put it in `/opt/gstreamer/lib/x86_64-linux-gnu`. Eliminated the whole CUDA runtime with this.

**Comment by ehfd** [comment]
Just an FYI: You'll like the latest commits. @remram44
